src/system_compositor.cpp

```diff
@@ -99,7 +123,24 @@ void usc::SystemCompositor::run()
                     mir::AbnormalExit("Video driver is blacklisted, exiting"));
             }
 
-            dm_connection = server->the_dm_connection();
+            if (server->get_options()->is_set("from-dm-fd") &&
+                server->get_options()->is_set("to-dm-fd"))
+            {
+                dm_connection = std::make_shared<usc::AsioDMConnection>(
+                    server->get_options()->get("from-dm-fd", -1),
+                    server->get_options()->get("to-dm-fd", -1),
+                    session_switcher);
+            }
+            else if (server->get_options()->is_set("debug-without-dm"))
+            {
+                dm_connection = std::make_shared<NullDMMessageHandler>(
```

Comment:
... so, in fact, it is that `session_switcher` is not initialized when the `NullDMMessageHandler` instance is created.  Rather than the order of the callbacks, my suspicion is that `session_switcher` is not initialized when the `SystemCompositor` constructor is called.

This suggests that the trick here is to provide `SystemCompositor` with a `the_session_switcher` delegate, but to keep `Server` as it is.  Should be straightforward enough in the circumstances.